### DISCUSSION TITLE
authExchange and NextJs

### DISCUSSION THREAD

**Discussion by bkinsey808** [urql-graphql/urql#2021] [Q&A] [answered]
It seems like when I include authExchange into my list of exchanges used by `withUrqlClient` I lose SSR with next. I can add other exchanges without break next ssr (e.g. urql devtools). But adding authExchange always breaks next's ssr. What's very weird is if I try to conditionally only add authExchange (`typeof window !== 'undefined'`) it still breaks ssr.

**Comment by JoviDeCroock** [comment]
It breaks doesn't tell us a lot 😅 in what way does it break? Could you reproduce it, ...

**Reply by bkinsey808** [reply]
I would like to create a smaller example, for now the code I have is here https://github.com/bkinsey808/nx-next-strapi particularly look at this file please: https://github.com/bkinsey808/nx-next-strapi/blob/main/apps/client/helpers/getUrqlClientOptions.ts

Anything obvious I'm missing or doing wrong?

When I say ssr "breaks" what I mean is that my index page is not generated with Next's SSR in production mode. With the authExchange in the array of exchanges, it breaks. Without, it's fine. Is there any example of authExchange used in a NextJs app I can look at?

**Reply by JoviDeCroock** [reply]
Looks good, seems like you're just not using tokens for the SSR which improves cacheability so the error must be related to not using auth I guess 😅 I have done a few bigger apps that used next and the auth-exchange and haven't noticed any issues so might be good to trace where the error is coming from.

We have this example which should be portable to Next https://github.com/FormidableLabs/urql/blob/main/examples/with-refresh-auth/src/client.js#L33 but generally I'd say if you're storing tokens in localStorage then it won't ever work on the server, cookies are more suited for this if you're looking for that

**Reply by bkinsey808** [reply]
True, it was intentional to not use tokens for ssr, the pages i want to ssr will not need authorization.

**Reply by bkinsey808** [reply]
I would love to look at some nextjs apps that use authExchange if source code is available for any of them...

**Reply by JoviDeCroock** [reply]
There are some NextJS repo's in here https://github.com/FormidableLabs/urql/network/dependents?package_id=UGFja2FnZS0xNDI2OTg4OTAx that use auth-exchange

**Reply by bkinsey808** [reply]
I'm really not fixated on storing the token in localstorage. Whatever works best. But, like i said, my reasoning is that I won't need authorization on the server side, I only care about SSR'ing pages with queries that don't need authorization.

**Reply by JoviDeCroock** [reply]
Mainly spitballing but have you tried using the `PostsDocument` in your index page rather than recreating the query twice? If your `html` generates correctly but the query doesn't match on the client that could be a reason for "breaks" but as I said it's hard to know whether you mean it errors or it missmatches or....

**Reply by bkinsey808** [reply]
No errors, it just doesn't SSR in production mode like it should when I include authExchange in the list of exchanges. SSR seems to work perfectly with no authExchange. Not sure what you mean about recreating the query twice. I was trying to follow the NextJs/Urql example here reasonably closely: https://formidable.com/open-source/urql/docs/advanced/server-side-rendering/#nextjs

**Reply by JoviDeCroock** [reply]
Try logging `console.log(result)` in `pages/index.tsx` with the result of `usePostsQuery`, what does it say? Also instead of recreating your query do `import { usePostsQuery, PostsDocument } from '../graphql';` and see if that helps as it might have something  to do with different hashes.

I really can't get that repo up and running 😅

**Reply by bkinsey808** [reply]
data looks like this: `{"posts":[{"id":"1","Title":" My First Blog Post","Content":"This is my first blog post. **This sentence is bold.** *This is italic.*\n\n- this \n- is\n- a\n- list\n\n","Slug":"my-first-blog-post","__typename":"Post"},{"id":"2","Title":"Second Post","Content":"This is a second post\n\n```ts\n const x = 1;\n```","Slug":"second-post","__typename":"Post"}]}` which is exactly what I would expect. Sorry, it's a hard repo to set up because you also have to get strapi up and running. I'll see if maybe I can repro on a smaller pure nextjs repo. Or maybe I'll just build two sets of urql client options: one for ssr with no authorization, and one for pages without ssr that has it.

**Reply by bkinsey808** [reply]
Also I should mention the entire `../graphql` file is codegen'd. That's why the query is written in the human code, but the usePostsQuery is imported from the codegen'd code. See https://www.graphql-code-generator.com/docs/plugins/typescript-urql

**Comment by bkinsey808** [comment, marked as answer]
The "solution" I have landed on for now is to decide on a page by page basis if I want authorization or not. Pages that require authorization will not have ssr. In code, my workaround solution looks like this:
```
export const withAuthUrqlClient = withUrqlClient(getUrqlClientOptions(true), {
  ssr: false,
});

export const withNoAuthUrqlClient = withUrqlClient(
  getUrqlClientOptions(false),
  {
    ssr: false,
  }
);
```
and
```
export const getUrqlClientOptions =
  (withAuth: boolean) =>
  (ssrCache: SSRExchange = ssrExchange({ isClient: true })) => ({
    url: process.env.NEXT_PUBLIC_CMS_GRAPHQL,
    exchanges: [
      devtoolsExchange,
      dedupExchange,
      cacheExchange,
      // for some reason, authExchange cannot be used on any page that should be SSR'd
      ...(withAuth
        ? [authExchange({ getAuth, addAuthToOperation, didAuthError })]
        : []),
      ssrCache,
      fetchExchange,
    ],
  });
```
